### ISSUE TITLE
One color per multiple labels results in gradients

### ISSUE THREAD

**Issue by vzinche** [napari/napari#1519]
## 🐛 (Probably a) Bug

Related to #716 

## Problem 
I have a segmentation image, which I want to color by some metadata. Let's say, each segment id has a property A or B assigned. To color segments by properties, I supply a dictionary to the labels layer that looks like:
`color_dict =  {0: 'black', 1: 'red', 2: 'red', 3: 'blue', 4: 'blue', etc}`
`viewer.add_labels(segmentation, color=color_dict)`
Note that in comparison to #716, I have one color per multiple labels. 

That is a sample resulting image I get:
![napari_color](https://user-images.githubusercontent.com/32451366/89294702-2c7fd100-d660-11ea-95a4-1aa350079a2d.png)

As you can notice, there are gradients of red and blue present (the purple cells). I've tried the very same with Image layer and vispy Colormap as described [here](https://forum.image.sc/t/specify-labels-colors/31534) and I still get the gradients. Where do they come from? It is not such a big issue for two colors, but once I have more it gets worse. 

## Expected behavior
Only blue, red and black colors on the image. No gradients

## Minimal reproducible example

```
def get_weird_gradients():
    sample_array = np.arange(100)
    np.random.shuffle(sample_array)
    labels_image = sample_array.reshape(10,10)
    colors = ['black', 'red', 'blue']

    color_map = {i: colors[np.random.randint(3)]
                 for i in range(100)}

    with napari.gui_qt():
        viewer = napari.Viewer()
        viewer.add_labels(labels_image, color=color_map) 
```

## Environment

napari: 0.3.6
Platform: Linux-4.17.9-041709-generic-x86_64-with-debian-buster-sid
Python: 3.7.6 | packaged by conda-forge | (default, Jun 1 2020, 18:57:50) [GCC 7.5.0]
Qt: 5.12.5
PyQt5: 5.12.3
NumPy: 1.18.4
SciPy: 1.4.1
Dask: 2.17.2
VisPy: 0.6.4

GL version: 4.6.0 NVIDIA 440.100
MAX_TEXTURE_SIZE: 32768

Plugins:
- napari-plugin-engine: 0.1.5
- svg: 0.1.3

## Additional context
Alternatively, is there any better way to color by metadata? 

cc @constantinpape

**Comment by constantinpape** [comment]
> Alternatively, is there any better way to color by metadata?

For some context here, @vzinche and me have been looking into similar problems of coloring labels by data associated with the labels.

I think a good solution for this would be to color via the `properties`, which are also available in the labels layer now. As far as I can see there's no option to explicitly color by the properties right now.
See also #1516.

**Comment by sofroniewn** [comment]
Thanks so much for flagging this, it does look like a bug from #1339 - maybe @ziyangczi can take a look.

It could either be in the colormap calculation here, https://github.com/napari/napari/blob/master/napari/utils/colormaps/colormaps.py#L119-L140 or in the mapping of integer labels to floats here https://github.com/napari/napari/blob/master/napari/layers/labels/labels.py#L556-L565. Somewhere things are getting out of sync.

It will be good to figure out where and add a test.

> I think a good solution for this would be to color via the properties, which are also available in the labels layer now. As far as I can see there's no option to explicitly color by the properties right now.

I agree. I don't think we've supported that yet, but that was definitely one of the goals and probably the long term better approach. There was some discussion of that here by @kevinyamauchi https://github.com/napari/napari/pull/1339#issuecomment-639650482

**Comment by liu-ziyang** [comment]
should be fixed with https://github.com/napari/napari/pull/1537, feel free to try it out!

**Comment by sofroniewn** [comment]
Closed by #1537 @constantinpape @vzinche please re-open if this isn't fixed as expected.
